staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go

```diff
@@ -212,51 +275,126 @@ func Start(nodeServer interface{}, opts ...Option) (result DRAPlugin, finalErr e
+		kubeClient: o.kubeClient,
+	}
+
+	// Stop calls cancel and therefore both cancellation
+	// and Stop cause goroutines to stop.
+	ctx, cancel := context.WithCancelCause(ctx)
+	d.backgroundCtx, d.cancel = ctx, cancel
+	logger.V(3).Info("Starting")
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		defer logger.V(3).Info("Stopping")
+		<-ctx.Done()
+	}()
+
+	// Clean up if we don't finish succcessfully.
+	defer func() {
+		if r := recover(); r != nil {
+			d.Stop()
+			panic(r)
+		}
+		if finalErr != nil {
+			d.Stop()
+		}
+	}()
+
 	// Run the node plugin gRPC server first to ensure that it is ready.
 	implemented := false
-	plugin, err := startGRPCServer(klog.LoggerWithName(o.logger, "dra"), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.draEndpoint, func(grpcServer *grpc.Server) {
-		if nodeServer, ok := nodeServer.(drapbv1alpha3.NodeServer); ok && o.nodeV1alpha3 {
-			o.logger.V(5).Info("registering drapbv1alpha3.NodeServer")
-			drapbv1alpha3.RegisterNodeServer(grpcServer, nodeServer)
+	plugin, err := startGRPCServer(klog.NewContext(ctx, klog.LoggerWithName(logger, "dra")), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.draEndpoint, func(grpcServer *grpc.Server) {
+		if nodeServer, ok := nodeServer.(drapb.NodeServer); ok && o.nodeV1alpha3 {
+			logger.V(5).Info("registering drapbv1alpha3.NodeServer")
+			drapb.RegisterNodeServer(grpcServer, nodeServer)
 			implemented = true
 		}
 	})
 	if err != nil {
 		return nil, fmt.Errorf("start node client: %v", err)
 	}
 	d.plugin = plugin
-	defer func() {
-		// Clean up if we didn't finish succcessfully.
-		if r := recover(); r != nil {
-			plugin.stop()
-			panic(r)
-		}
-		if finalErr != nil {
-			plugin.stop()
-		}
-	}()
 	if !implemented {
 		return nil, errors.New("no supported DRA gRPC API is implemented and enabled")
 	}
 
 	// Now make it available to kubelet.
-	registrar, err := startRegistrar(klog.LoggerWithName(o.logger, "registrar"), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, o.draAddress, o.pluginRegistrationEndpoint)
+	registrar, err := startRegistrar(klog.NewContext(ctx, klog.LoggerWithName(logger, "registrar")), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, o.draAddress, o.pluginRegistrationEndpoint)
 	if err != nil {
 		return nil, fmt.Errorf("start registrar: %v", err)
 	}
 	d.registrar = registrar
 
+	// startGRPCServer and startRegistrar don't implement cancellation
+	// themselves, we add that for both here.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		<-ctx.Done()
+
+		// Time to stop.
+		d.plugin.stop()
+		d.registrar.stop()
+
+		// d.resourceSliceController is set concurrently.
+		// Stop may get called with nil receiver, so we
+		// don't need to check here.
+		d.mutex.Lock()
+		d.resourceSliceController.Stop()
+		d.mutex.Unlock()
+
+		// The stop calls have blocked. Now that we are done,
+		// d.wg.Done() above can be called to mark everything as
+		// stopped.
```

Comment:
Is this comment really necesssary?

Reply:
If you didn't find it useful, then not... removed. :smile: